Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1636,22 +1668,141 @@ void CGContextSetCMYKFillColor(CGContextRef context, CGFloat cyan, CGFloat magen
+
+    CGContextSetFillColorSpace(patternContext.get(), context->fillColorSpace());
+    // set the fill color
+    CGContextSetFillColor(patternContext.get(), components);
+
+    // Now we ask the user to draw
+    _CGPaternIssueCallBack(patternContext.get(), pattern);
+
+    // Get the image out of it
+    woc::unique_cf<CGImageRef> tileImage{ __CGContextCreateRenderableImage(CGBitmapContextCreateImage(patternContext.get())) };
+    RETURN_HR_IF_NULL(E_UNEXPECTED, tileImage);
+
+    ComPtr<ID2D1Bitmap> d2dBitmap;
+    RETURN_IF_FAILED(__CreateD2DBitmapFromCGImage(context, tileImage.get(), &d2dBitmap));
+
+    // Scale it by the inverted transform
+    // TODO: We have an issue with rotation, the CTM rotation should not affect the brush
+    CGSize size = CGSizeApplyAffineTransform(tileSize.size, CGAffineTransformInvert(context->CurrentGState().transform));
+
+    CGAffineTransform transform = __BitmapBrushTransformation(context,
+                                                              { { 0, 0 }, size.width, size.height },
+                                                              d2dBitmap->GetPixelSize(),
+                                                              _CGPatternGetTransformation(pattern));
+
+    ComPtr<ID2D1BitmapBrush1> bitmapBrush;
+    ComPtr<ID2D1DeviceContext> deviceContext = context->DeviceContext();
+    RETURN_IF_FAILED(deviceContext->CreateBitmapBrush(d2dBitmap.Get(),
+                                                      D2D1::BitmapBrushProperties1(D2D1_EXTEND_MODE_WRAP,
+                                                                                   D2D1_EXTEND_MODE_WRAP,
+                                                                                   context->CurrentGState().GetInterpolationModeForCGImage(
+                                                                                       tileImage.get())),
+                                                      D2D1::BrushProperties(1 /* the global alpha will be set during the draw sequence*/,
+                                                                            __CGAffineTransformToD2D_F(transform)),
+                                                      &bitmapBrush));
+
+    *brush = bitmapBrush.Detach();
+    return S_OK;
+}
+
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
-void CGContextSetStrokePattern(CGContextRef context, CGPatternRef pattern, const CGFloat* components) {
+void CGContextSetFillPattern(CGContextRef context, CGPatternRef pattern, const CGFloat* components) {
     NOISY_RETURN_IF_NULL(context);
-    UNIMPLEMENTED();
+    NOISY_RETURN_IF_NULL(pattern);
+    NOISY_RETURN_IF_NULL(components);
+
+    ComPtr<ID2D1BitmapBrush1> bitmapBrush;
+    FAIL_FAST_IF_FAILED(_CreatePatternBrush(context, pattern, components, &bitmapBrush));
+    // set the fill brush
+    FAIL_FAST_IF_FAILED(bitmapBrush.As(&context->CurrentGState().fillBrush));
 }
 
 /**
  @Status Interoperable
```

Comment:
"The number of array elements must equal the number of components in the base space of the stroke pattern color space,"

Pretty much it should be set.

---
In reply to: [92969524](https://github.com/Microsoft/WinObjC/pull/1586#discussion_r92969524) [](ancestors = 92969524)